sdk/servicebus/service-bus/src/serviceBusClient.ts

```diff
@@ -15,6 +15,7 @@ import { CreateSessionReceiverOptions, CreateSenderOptions } from "./models";
 import { Receiver, ReceiverImpl } from "./receivers/receiver";
 import { SessionReceiver, SessionReceiverImpl } from "./receivers/sessionReceiver";
 import { ReceivedMessageWithLock, ReceivedMessage } from "./serviceBusMessage";
+import { openLink } from "./shared/openLink";
```

Comment:
They could - this is to make unit testing possible by always injecting openLink.

Reply:
Is injecting the only way to ensure unit testability? My concern is that this model is not intuitive and adds complexity that passing around a function brings.

If `openLink` was meant to be a common shared code between `MessageSender` and `MessageReceiver`, then I would expect it to be part of `LinkEntity` which is the base class for both. An independent helper method makes sense when it is being used from multiple unrelated places & is stateless. Given that this helper is only ever going to be used by classes that inherit from `LinkEntity`, the base class is a much better fit

Seeing the unit tests, I mainly see the below categories
- Ensure abort signal is plumbed through to open()
- Ensure open() is not called when abort signal is already fired

If this common code was on `LinkEntity`, then for the above unit tests, we can override the method for open(). I see that the plumbing needs to be tested from SenderImpl -> MessageSender -> LinkEntity. Would perhaps combining SenderImpl & MessageSender help?

`export class SenderImpl extends LinkEntity implements Sender` ?

Then you test code would be
```js
class TestAbortSignalSender extends SenderImpl {
     async openLink(..) {
        if (options.abortSignal != null) {
          throw new Error("abortSignal was properly passed to open()!");
        } else {
          throw new Error("No abortSignal was passed to open()!");
        }
     } 
}
```
instead of 
```js
sender = new SenderImpl(clientEntityContext, {}, async (openArgs) => {
       if (openArgs.abortSignal != null) {
          throw new Error("abortSignal was properly passed to open()!");
        } else {
          throw new Error("No abortSignal was passed to open()!");
        }
});
```

cc @chradek

Reply:
Definitely not the only way. There are so many ways!

Inheritance for testing is also a valid approach and I don't think it's bad. In the code as it is I think inheritance is making it difficult to see where the lines are between what's mutable, and what's not.

Also, this is definitely not this code's final form but it is a step in the direction that I think helps with making the code easier to maintain. Some state is ultimately required (so we might want to consider how to do that) though but I didn't want to do too much in one PR.